### ISSUE TITLE
Build problem with htslib

### ISSUE THREAD

**Issue by evolgenomics** [wtsi-npg/bambi#170]
I have an issue with htslib. I've tried now with htslib v1.3.1, 1.4, 1.7, 1.8 and 1.9 - most everything gives the same error. 

Am I missing something?

Thanks!

```make  all-am
make[1]: Entering directory '/fml/chones/local/bambi'
gcc -DHAVE_CONFIG_H -I.  -I./src -I/fml/chones/local/include  -std=gnu99  -I/fml/chones/local/include/libxml2 -std=gnu99 -Wall -g -O2 -Werror -c -o src/decode.o src/decode.c
In file included from src/decode.c:42:0:
src/bamit.h:33:5: error: unknown type name ‘sam_hdr_t’
     sam_hdr_t *h;
     ^~~~~~~~~
src/bamit.h:52:33: error: unknown type name ‘sam_hdr_t’; did you mean ‘bam_hdr_t’?
 BAMit_t *BAMit_init(samFile *f, sam_hdr_t *h);
                                 ^~~~~~~~~
                                 bam_hdr_t
src/decode.c:1025:22: error: unknown type name ‘sam_hdr_t’; did you mean ‘bam_hdr_t’?
 static void addNewRG(sam_hdr_t *hdr, char *entry, char *bcname, char *lib, char *sample, char *desc)
                      ^~~~~~~~~
                      bam_hdr_t
src/decode.c:1084:46: error: unknown type name ‘sam_hdr_t’; did you mean ‘bam_hdr_t’?
 static void changeHeader(va_t *barcodeArray, sam_hdr_t *sh, char *argv_list)
                                              ^~~~~~~~~
                                              bam_hdr_t
src/decode.c: In function ‘processTemplatesNoThreads’:
src/decode.c:1273:44: error: passing argument 2 of ‘sam_write1’ from incompatible pointer type [-Werror=incompatible-pointer-types]
             int r = sam_write1(bam_out->f, bam_out->h, rec_n);
                                            ^~~~~~~
In file included from src/decode.c:23:0:
/fml/chones/local/include/htslib/sam.h:394:9: note: expected ‘const bam_hdr_t * {aka const struct <anonymous> *}’ but argument is of type ‘int *’
     int sam_write1(samFile *fp, const bam_hdr_t *h, const bam1_t *b) HTS_RESULT_USED;
         ^~~~~~~~~~
src/decode.c: In function ‘output_job_results’:
src/decode.c:1325:40: error: passing argument 2 of ‘sam_write1’ from incompatible pointer type [-Werror=incompatible-pointer-types]
         int r = sam_write1(bam_out->f, bam_out->h, rec);
                                        ^~~~~~~
In file included from src/decode.c:23:0:
/fml/chones/local/include/htslib/sam.h:394:9: note: expected ‘const bam_hdr_t * {aka const struct <anonymous> *}’ but argument is of type ‘int *’
     int sam_write1(samFile *fp, const bam_hdr_t *h, const bam1_t *b) HTS_RESULT_USED;
         ^~~~~~~~~~
src/decode.c: In function ‘decode’:
src/decode.c:1641:9: error: implicit declaration of function ‘sam_hdr_destroy’; did you mean ‘bam_hdr_destroy’? [-Werror=implicit-function-declaration]
         sam_hdr_destroy(bam_out->h); bam_out->h = sam_hdr_dup(bam_in->h);
         ^~~~~~~~~~~~~~~
         bam_hdr_destroy
src/decode.c:1641:51: error: implicit declaration of function ‘sam_hdr_dup’; did you mean ‘bam_hdr_dup’? [-Werror=implicit-function-declaration]
         sam_hdr_destroy(bam_out->h); bam_out->h = sam_hdr_dup(bam_in->h);
                                                   ^~~~~~~~~~~
                                                   bam_hdr_dup
src/decode.c:1641:49: error: assignment makes pointer from integer without a cast [-Werror=int-conversion]
         sam_hdr_destroy(bam_out->h); bam_out->h = sam_hdr_dup(bam_in->h);
                                                 ^
src/decode.c:1644:9: error: implicit declaration of function ‘changeHeader’; did you mean ‘print_header’? [-Werror=implicit-function-declaration]
         changeHeader(barcodeArray, bam_out->h, opts->argv_list);
         ^~~~~~~~~~~~
         print_header
src/decode.c:1645:39: error: passing argument 2 of ‘sam_hdr_write’ from incompatible pointer type [-Werror=incompatible-pointer-types]
         if (sam_hdr_write(bam_out->f, bam_out->h) != 0) {
                                       ^~~~~~~
In file included from src/decode.c:23:0:
/fml/chones/local/include/htslib/sam.h:384:9: note: expected ‘const bam_hdr_t * {aka const struct <anonymous> *}’ but argument is of type ‘int *’
     int sam_hdr_write(samFile *fp, const bam_hdr_t *h) HTS_RESULT_USED;
         ^~~~~~~~~~~~~
cc1: all warnings being treated as errors
Makefile:1082: recipe for target 'src/decode.o' failed
make[1]: *** [src/decode.o] Error 1
make[1]: Leaving directory '/fml/chones/local/bambi'
Makefile:806: recipe for target 'all' failed
```make: *** [all] Error 2

**Comment by jenniferliddle** [comment]
Ah. The problem is that the missing functions have only appeared since htslib version 1.10 (we're now on version 1.12).
It looks like I've forgotten to update the check for htslib version in the bambi configure script.
Thanks for raising the issue. I'll fix this.

**Comment by evolgenomics** [comment]
Right - didn't realize it's been a while since I pulled htslib fresh. 

Replaced it and have gotten past the problem. Still, there's some build issues with `libz` now?

I'm compiling it not as root - but replacing the various libz dynlib didn't seem to help. Our `zlib` version is v1.2.11.

```frankyc@congo:/fml/chones/local/bambi 02:16 PM : make
make  all-am
make[1]: Entering directory '/fml/chones/local/bambi'
/bin/bash ./libtool  --tag=CC   --mode=link gcc -I/fml/chones/local/include/libxml2 -std=gnu99 -Wall -g -O2 -Werror -rdynamic   -o src/bambi src/bambi.o src/bambi_utils.o src/decode.o src/i2b.o src/select.o src/chrsplit.o src/read2tags.o src/spatial_filter.o src/seqchksum.o src/adapters.o src/update.o src/bclfile.o src/filterfile.o src/posfile.o src/array.o src/parse.o src/bamit.o src/hash_table.o src/parse_bam.o src/crc.o  -lm -lxml2 
libtool: link: gcc -I/fml/chones/local/include/libxml2 -std=gnu99 -Wall -g -O2 -Werror -rdynamic -o src/bambi src/bambi.o src/bambi_utils.o src/decode.o src/i2b.o src/select.o src/chrsplit.o src/read2tags.o src/spatial_filter.o src/seqchksum.o src/adapters.o src/update.o src/bclfile.o src/filterfile.o src/posfile.o src/array.o src/parse.o src/bamit.o src/hash_table.o src/parse_bam.o src/crc.o  -lm -lxml2
/usr/bin/ld: src/bclfile.o: undefined reference to symbol 'inflateEnd'
//lib/x86_64-linux-gnu/libz.so.1: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
Makefile:941: recipe for target 'src/bambi' failed
make[1]: *** [src/bambi] Error 1
make[1]: Leaving directory '/fml/chones/local/bambi'
Makefile:806: recipe for target 'all' failed
make: *** [all] Error 2

```frankyc@congo:/fml/chones/local/bambi 02:16 PM : ls /lib/x86_64-linux-gnu/libz.so.1
/lib/x86_64-linux-gnu/libz.so.1
frankyc@congo:/fml/chones/local/bambi 02:16 PM : ls /lib/x86_64-linux-gnu/libz.so*
/lib/x86_64-linux-gnu/libz.so.1  /lib/x86_64-linux-gnu/libz.so.1.2.11
```

**Comment by jenniferliddle** [comment]
When you ran ./configure did you specify a --with-htslib option? If not, can you try that?
It looks as if some of the libraries (such as -lz) are not being specified, and explicitly setting --with-htslib may be a workaround.

**Comment by srl147** [comment]
Jennifer

the pb_calibration configure.ac script doesn't understand --with-hstlib option, I checked the bambi version but I'm not sure what I should be adding

regards
Steven
________________________________________

**Comment by evolgenomics** [comment]
I have also played around with various different options there. In my case I suspect it's due to some non-standard install locations, etc, so my problem may not be general. I plan to dig into this a bit more - hopefully I can resolve this myself and close the issue. 

I did experience the same as Steven/srl147 that the --with-htslib option is a bit dysfunctional/misleading. I end up having to specify either by modifying the Makefile or by declaring HTSLIB_HOME up-front as an environmental variable.
